aorsf learner

bcjaeger opened the pull request:
# New Learner

This pull request adds **LearnerSurvAorsf** with the unique key **surv.aorsf** from package **{aorsf}**.

I have...

* [X] Created a learner with the create_learner template or created one manually and it follows the identical layout.
* [X] Added `.train` and `.predict` private methods
* [X] Added `ParamSet` and if applicable `param_vals`
* [X] Added `importance` public method and completed the documentation for this - !!!Delete if not applicable!!!
* [X] Added `oob_error` public method and completed the documentation for this - !!!Delete if not applicable!!!
* [X] Added references and not left the tag blank - !!!Delete if not applicable!!!
* [X] Removed all 'FIXME's

## Dependencies

I have...

* [ ] Added more than one dependency to suggests - If ticked, explain why below.
* [X] Added no dependencies to imports.
* [ ] Added dependencies to remotes - If ticked, explain why below.

I have added `{aorsf}` to suggests

## Tests

I have...

* [X] Created a testthat file and it passes locally.
* [X] Created a paramtest file with tests for all calls in both my `.train` and `.predict` methods and it passes locally.

## Cleaning

I have run...

* [X] `devtools::document(roclets = c('rd', 'collate', 'namespace'))`
* [X] `lintr::lint_package()`

And

* [X] Updated NEWS.md

## Maintenance and Contributions

This PR will only be merged if the following is filled in and ticked:

* [X] bcjaeger agrees to maintain this learner going forward and agrees that this learner will be removed from `mlr3extralearners` if: i) the build remains broken for more than one month with no effort to fix it; or ii) is broken for more than three months.

The maintainer of the learner can be changed at any time by editing the '@author' tag and creating a pull request.

Finally make sure you get credit and add yourself to the DESCRIPTION file (optionally add a note to say which learner(s) you implemented):

* [X] I have added myself (and anyone else who contributed) to the end of the "Authors" field in the DESCRIPTION file with role "ctb"

## Comments

Hello! I am following up on #198. `aorsf` has undergone review at rOpenSci<sup>1</sup> and is stable. It was also recently uploaded to CRAN. 

<sup>1</sup>Technically, the rOpenSci issue is still open (https://github.com/ropensci/software-review/issues/532), but I anticipate the editor will be closing it sometime this week.

sebffischer commented:
Thank you for contributing!

To keep you updated: 
I wanted to merge some small fixes first. I will update your branch and review this PR hopefully later today

bcjaeger commented:
Awesome! Thanks for letting me know - I'm happy to make updates based on your review.

sebffischer reviewed (commented):
Looking good! Most of my comments are minor details (where we should have documented better) or styling. Don't hesitate to ask if something is unclear! 

And thanks again for contributing!

Do you think we should add a warning when using your slower implementation to make users aware that a faster version exists?

sebffischer commented:
Also you need to add bcjaeger/aorsf to Remotes in the DESCRIPTION because it is not available on CRAN it seems

bcjaeger commented:
Thank you for reviewing so carefully - the suggested edits are great! 

I will check out and resolve each of the conversations above, and let you know when I've gotten everything wrapped up.

sebffischer commented:
Great. 
For some reason it cannot find the aorsf but I am not sure. Will check tomorrow.

bcjaeger commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,278 @@
+            #' OOB concordance error extracted from the model slot `oob_error`.
+            #' @return `numeric()`.
+            oob_error = function() {
+
```
Got it! I left empty lines between the public and private lists and also empty lines between items in those lists. Can drop more if that is helpful.

bcjaeger commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,278 @@
+              ps$control_net_alpha <- NULL
+              ps$control_net_df_target <- NULL
+
+              tn = task$target_names
```
Seems like it works alright for survival tasks, great suggestion!

bcjaeger commented on tests/testthat/test_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,13 @@
+
+test_that("autotest", {
+  set.seed(1) # there was presumable a weird bug
```
Ahh, sorry about that.  I copied and pasted this script from `test_obliqueRSF_surv_obliqueRSF` and forgot to delete this comment - my mistake! I do not know what the bug was for `obliqueRSF`, but I did not encounter a bug while setting up the tests for `aorsf`, so this comment will be deleted.

bcjaeger commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,278 @@
+            },
+
+            #' @description
+            #' OOB concordance error extracted from the model slot `oob_error`.
```
My mistake - changed this from `oob_error` to `eval_oobag$stat_values`

bcjaeger commented:
> Do you think we should add a warning when using your slower implementation to make users aware that a faster version exists?

This is a great idea - should I put a `warning()` about this in the initialize function for obliqueRSF learners?

bcjaeger commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,278 @@
+                                 ratio = "mtry_ratio",
+                                 length(task$feature_names))
+
+              ps = c(ps, list(weights = task$weights$weight))
```
Good point - I realized I could plug `task$weights` into `aorsf::orsf()`, which allows `weights = NULL`

bcjaeger commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,278 @@
+            .train = function(task) {
+
+              # start w/defaults
+              ps <- self$param_set$default
```
Noted! I've initialized following the example linked and added a helper function to `.train` so that I can organize the input for `aorsf::fit()` and get the correct type of `control` object. In retrospect making the `control` input a function was not a good choice for compatibility with `mlr3`, whoops =]

bcjaeger commented:
`aorsf` should be on CRAN (https://cran.r-project.org/web/packages/aorsf/index.html) so I haven't added the remotes part yet. But if the CRAN link still doesn't work I am happy to add a remote.

sebffischer commented:
> > Do you think we should add a warning when using your slower implementation to make users aware that a faster version exists?
> 
> This is a great idea - should I put a `warning()` about this in the initialize function for obliqueRSF learners?

Yes I would do it like that.

sebffischer commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,182 @@
+            #' Creates a new instance of this [R6][R6::R6Class] class.
+            initialize = function() {
+              ps = ps(
+                n_tree = p_int(default = 500L, lower = 1L, tags = 'train'),
```
some single quotes are still left, please replace with ""

bcjaeger replied:
Sorry about that, should be fixed now!

sebffischer commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,182 @@
+          ),
+
+          private = list(
+            .train = function(task) {
```
Please replace `<-` with `=`, I know this is not typical R, but we prefer to have it consistent.

bcjaeger replied:
`<-` should be used everywhere now, apart from the places where only `=` works.

sebffischer commented on R/learner_aorsf_surv_aorsf.R:
```diff
@@ -0,0 +1,182 @@
+                control_type = p_fct(levels = c("fast", "cph", "net"),
+                                     default = "fast", tags = "train"),
+                control_fast_do_scale = p_lgl(default = TRUE, tags = "train"),
+                control_fast_method = p_fct(levels = c("efron", "breslow"),
```
here the default is set to "breslow", but when I look at the docs from orsf_control_fast, the default is "efron". 
Please document this in the learner documentation under the section "custom mlr3 defaults" like here: https://github.com/mlr-org/mlr3extralearners/blob/56d38b1401bad469631aa912101d659caa3ad536/R/learner_lightgbm_regr_lightgbm.R#L26

Please also explain these custom parameters that are not actually documented in the aorsf package (like control_fast_do_scale). Otherwise one has to look at the code, or guess what they do

sebffischer replied:
Please also check this for the other custom parameters (i.e. parameters whose name is not identical to the name of a parameter in the functions from the aorsf package)

bcjaeger replied:
My mistake - "efron" was the method I intended to put as the default here. After reviewing other inputs for `orsf` and the control functions, I found no other discrepancies between default values.

sebffischer reviewed (commented):
Thanks for fixing most of the issues! 
I have left some more comments.

sebffischer commented:
I think the problem with the installation is in the CI and not with your package

sebffischer commented:
So your tests are passing! 
The errors are still problems with the CI that I have to address, but when you address the issues I raised your PR looks good to be merged :)

bcjaeger commented:
I think I've addressed all of the comments now. 

One thing I am trying to figure out: I think `predict.orsf_fit()` is exported from `aorsf`, but when I use `aorsf::predict.orsf_fit()` in `run_paramtest()` I get an error saying `predict.orsf_fit()` is not exported. Codefactor does not like me using `:::`, so I've modified the `paramtest` file for `aorsf`, using `predict()` in `fun_list` instead of `aorsf:::predict.orsf_fit()`. The test passes for me locally with this workaround, and I don't think it upsets codefactor. What do you think?

sebffischer commented:
Sorry for the late reply. Your workaroud is fine. In the tests I think codefactor issues can be ignore, exactly for the reason you mention.

sebffischer commented:
I formatted your code using the mlr3 style, which addressed the only comments that I would have made. 

Thanks for contributing and for addressing all my comments ! :)

sebffischer reviewed (commented):
lgtm!

bcjaeger commented:
Awesome - thank you! =]
